documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,239 @@
+
+The implicit project file is the default project that would be created by running `dotnet new console`.
+This means that the behavior of `dotnet run file.cs` can change between SDK versions if the `dotnet new console` template changes.
+In the future we can consider supporting more SDKs like the Web SDK.
+
+## Grow up
+
+When file-based programs reach an inflection point where build customizations in a project file are needed,
+a single CLI command can be executed to generate a project file.
+In fact, this command simply materializes the [implicit project file](#implicit-project-file) on disk.
+This action should not change the behavior of the target program.
+
+```ps1
+dotnet project add
+```
+
+## Target path
+
+The path passed to `dotnet run ./some/path.cs` is called *the target path*.
+If it is a file, it is called *the target file*.
+*The target directory* is the directory of the target file, or the target path if it is not a file.
+
+We can consider adding an option like `dotnet run --from-stdin` which would read the C# file from the standard input.
+In this case, the current working directory would not be used to search for project or other C# files,
+the compilation would consist solely of the single file read from the standard input.
+
+## Integration into the existing `dotnet run` command
+
+`dotnet run file.cs` already has a meaning if there is a project file inside the current directory,
+specifically `file.cs` is passed as the first command-line argument to the target program.
+We preserve this behavior to avoid a breaking change.
+The file-based build and run kicks in only when:
+- a project file cannot be found (in the current directory or via the `--project` option), and
+- if the target path is a file, it has the `.cs` file extension, and
+- the target path (file or directory) exists.
+
+> [!NOTE]
+> This means that `dotnet run path` stops working when a file-based program [grows up](#grow-up) into a project-based program.
+>
+> Users could avoid that by using `cd path; dotnet run` instead. For that to work always (before and after grow up),
+> `dotnet run` without a `--project` argument and without a project file in the current directory
+> would need to search for a file-based program in the current directory instead of failing.
+>
+> We can also consider adding some universal option that would work with both project-based and file-based programs,
+> like `dotnet run --path ./dir/`.
+
+File-based programs are processed by `dotnet run` equivalently to project-based programs unless specified otherwise in this document.
+For example, the remaining command-line arguments after the first argument (the target path) are passed through to the target app
+(except for the arguments recognized by `dotnet run` unless they are after the `--` separator).
+
+### Unsupported `dotnet run` options
+
+These `dotnet run` options are not supported by file-based programs (they result in an error):
+- `--launch-profile`
+- `--no-build`
```

Comment:
It wasn't clear to me why this option should not work, if an artifact was already built for the indicated source file.

Reply:
You're right, that should work eventually. It just doesn't work in the initial stage where we always build, although I can look if that could be also fixed easily.